### ISSUE TITLE
Publisher-Confirms over TLS sometimes take longer than expected

### ISSUE THREAD

**Issue by Keruspe** [sozu-proxy/lapin#267]
Let's reopen #251 with a clean state.

My next five comments here will be:
- the logs of a test program using plain amqp for ~1h
- the logs of a test program using amqps with openssl for ~1h
- the logs of a test program using amqps with native-tls for ~1h
- the logs of a test program using amqps with rustls for ~1h
- a relevant part the logs of a test program using amqps with rustls with RUST_LOG=trace

Couple of things about what to expect here:
- AFAICT the issue cannot be reproduced using plain amqp
- openssl and native-tls should be roughly the same as I'm running the tests on linux
- there is another known issue with native-tls where the stream claims to be writable but we can only write 0 bytes (half closed? why? why is it not detected by mio/epoll?). I added a tweak to consider we've hit EWOULDBLOCK when this happens
- after running for several hours, the rustls test last time hit this 24 times over 2500 messages.

The broker used for the tests has poor connectivity (to my taste at least) in ordre to reproduce this issue more easily as it seems to be dependent on the network lag. Typical ack time is around 14s, sometimes up to 16-17. Problematic response time are usualy between 60 and 70s.

**Comment by Keruspe** [comment]
Plain AMQP after roughly one hour:

```
[2020-04-16T07:28:03Z DEBUG lapin::channels] create channel with id 0
[2020-04-16T07:28:04Z DEBUG lapin::channel] Server sent Connection::Tune: Tune { channel_max: 200, frame_max: 131072, heartbeat: 120 }
[2020-04-16T07:28:05Z INFO  foo] CONNECTED
[2020-04-16T07:28:05Z DEBUG lapin::channels] create channel
[2020-04-16T07:28:05Z DEBUG lapin::channels] create channel with id 1
[2020-04-16T07:28:05Z INFO  foo] Enabled publisher-confirms
[2020-04-16T07:28:05Z INFO  foo] will publish
1432281 sent 100 messages
2559814 msg 179 took 16404ms to ack
2605196 msg 182 took 16948ms to ack
2861400 sent 200 messages
3065101 msg 214 took 16132ms to ack
```

3 messages with a lag a little bit slower that usual, but nothing close to the minute.

**Comment by Keruspe** [comment]
OpenSSL after a little more than 30 minutes:

```
[2020-04-16T08:28:51Z DEBUG lapin::channels] create channel with id 0
[2020-04-16T08:28:52Z DEBUG lapin::channel] Server sent Connection::Tune: Tune { channel_max: 200, frame_max: 131072, heartbeat: 120 }
[2020-04-16T08:28:52Z INFO  foo] CONNECTED
[2020-04-16T08:28:52Z DEBUG lapin::channels] create channel
[2020-04-16T08:28:52Z DEBUG lapin::channels] create channel with id 1
[2020-04-16T08:28:53Z INFO  foo] Enabled publisher-confirms
[2020-04-16T08:28:53Z INFO  foo] will publish
[2020-04-16T08:28:54Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:29:54Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:30:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:30:54Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:31:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:31:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:31:54Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:32:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:32:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:33:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:33:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:33:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:34:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:34:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:35:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:35:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:35:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:36:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:36:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:37:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:37:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:37:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:38:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:38:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:39:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:39:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:39:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:40:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:40:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:41:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:41:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:41:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:42:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:42:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:43:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:43:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:43:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:44:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:44:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:45:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:45:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:45:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:46:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:46:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:47:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:47:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:47:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:48:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:48:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:49:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:49:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:49:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:50:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:50:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:51:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:51:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:51:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:52:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:52:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:53:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:53:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:53:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:54:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:54:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:55:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:55:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:55:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:56:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:56:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:57:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:57:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:57:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T08:58:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:58:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:59:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T08:59:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T08:59:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T09:00:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T09:00:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T09:01:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T09:01:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T09:01:52Z DEBUG lapin::channels] send heartbeat
[2020-04-16T09:02:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T09:02:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T09:03:52Z ERROR lapin::io_loop] Socket was writable but we wrote 0, marking as wouldblock
[2020-04-16T09:03:52Z DEBUG lapin::channels] received heartbeat from server
[2020-04-16T09:03:52Z DEBUG lapin::channels] send heartbeat
```

I think we won't be able to properly test this with openssl/native-tls until we've figured out this other problem

**Comment by Keruspe** [comment]
as expected, native-tls gives the same output

**Comment by Keruspe** [comment]
Rustls:

```
[2020-04-16T09:11:45Z DEBUG lapin::channels] create channel with id 0
[2020-04-16T09:11:45Z DEBUG lapin::channel] Server sent Connection::Tune: Tune { channel_max: 200, frame_max: 131072, heartbeat: 120 }
[2020-04-16T09:11:46Z INFO  foo] CONNECTED
[2020-04-16T09:11:46Z DEBUG lapin::channels] create channel
[2020-04-16T09:11:46Z DEBUG lapin::channels] create channel with id 1
[2020-04-16T09:11:47Z INFO  foo] Enabled publisher-confirms
[2020-04-16T09:11:47Z INFO  foo] will publish
123660 msg 8 took 16634ms to ack
[2020-04-16T09:32:46Z DEBUG lapin::channels] received heartbeat from server
1262655 msg 82 took 62838ms to ack
[2020-04-16T09:38:11Z DEBUG lapin::channels] send heartbeat
1588065 msg 100 took 73422ms to ack
1588066 sent 100 messages
[2020-04-16T09:40:40Z DEBUG lapin::channels] send heartbeat
1737196 msg 106 took 73766ms to ack
1753358 msg 107 took 16161ms to ack
[2020-04-16T09:42:39Z DEBUG lapin::channels] send heartbeat
1855593 msg 110 took 73188ms to ack
```

**Comment by Keruspe** [comment]
Relevant trace logs with rustls

```
[2020-04-16T10:07:32Z TRACE lapin::io_loop] wrote 262144 bytes
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:32Z TRACE lapin::io_loop] Will write to buffer
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=false, has_data=true, status=Setup
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:32Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:07:33Z TRACE lapin::io_loop] Got mio event for socket: Event { token: Token(1), readable: false, writable: true, error: false, read_closed: false, write_closed: false, priority: false, aio: false, lio: false }
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:33Z TRACE lapin::io_loop] Will write to buffer
[2020-04-16T10:07:33Z TRACE lapin::io_loop] wrote 262144 bytes
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:33Z TRACE lapin::io_loop] Will write to buffer
[2020-04-16T10:07:33Z TRACE lapin::io_loop] wrote 97720 bytes
[2020-04-16T10:07:33Z TRACE lapin::io_loop] 33352 to write to complete this frame
[2020-04-16T10:07:33Z TRACE lapin::io_loop] Still 33352 to write
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-16T10:07:33Z TRACE lapin::io_loop] Will write to buffer
[2020-04-16T10:07:33Z TRACE lapin::io_loop] wrote 33744 bytes
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:33Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:07:34Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:07:34Z TRACE lapin::io_loop] Got mio event for socket: Event { token: Token(1), readable: false, writable: true, error: false, read_closed: false, write_closed: false, priority: false, aio: false, lio: false }
[2020-04-16T10:07:34Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=false
[2020-04-16T10:07:34Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-16T10:07:34Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:07:34Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=false
[2020-04-16T10:08:33Z DEBUG lapin::channels] send heartbeat
[2020-04-16T10:08:33Z TRACE lapin::channel] channel 0 send_frame
[2020-04-16T10:08:33Z TRACE lapin::channel] channel 0 wake
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-16T10:08:33Z TRACE lapin::io_loop] Will write to buffer
[2020-04-16T10:08:33Z TRACE lapin::io_loop] wrote 8 bytes
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop run
[2020-04-16T10:08:33Z TRACE lapin::io_loop] io_loop poll
[2020-04-16T10:08:34Z TRACE lapin::io_loop] io_loop poll done
[2020-04-16T10:08:34Z TRACE lapin::io_loop] Got mio event for socket: Event { token: Token(1), readable: true, writable: true, error: false, read_closed: false, write_closed: false, priority: false, aio: false, lio: false }
[2020-04-16T10:08:34Z TRACE lapin::io_loop] io_loop do_run; can_read=true, can_write=true, has_data=false
[2020-04-16T10:08:34Z TRACE lapin::io_loop] read 21 bytes
[2020-04-16T10:08:34Z TRACE lapin::channels] will handle frame: Method(1, Basic(Ack(Ack { delivery_tag: 74, multiple: false })))
1128090 msg 74 took 73611ms to ack
```

**Comment by Keruspe** [comment]
At some point, we write a bit less than the usual chucks of 262144 bytes, writing only 97720 bytes. At this moment, there's still 33352 serialized bytes to send. Next write is 33744, meaning we enqueued 392 more bytes in the meantime. Not sure what those 392 bytes are, worth investigating by logging which frames we serialize and when.

Once we're written those last 33744 bytes, we do not any more serialized data to send, and has_data is false so we don't have any frames pending serialization either.

One second after that, we get a very strange mio event. Given the heartbeat timeout is 60s here, and we just wrote stuff, it's not the timeout that causes this (worth adding logs to confirm though). The reason I find this event weird is that it doesn't carry any data beside the writable flag, when the socket was already writable anyways.

Socket is not readable, we have noting to send, so we don't do anything in this iteration, and we get back to poll().

This time, after 59s (60s since last write, as we got a spurious event 1s after last write), we hit the timeout used for heartbeating. The log for poll_done tells us that our state hasn't change, socket is still not readable and we still have nothing to do.

We finish this iteration by scheduling a heartbeat and go back to poll. Poll gets immediately awoken by the heartbeat waker, this time we have some data to send.

We write the 8 bytes corresponding to the heartbeat and magically, socket becomes readable and we can read our ack.

I'll add some logging of the actual data we write when writing small debugable amounts of data to ensure the 8 bytes we wrote are indeed the exact 8 bytes from the heartbeat.

**Comment by Keruspe** [comment]
Re ran this with the extra logs added:

The extra bytes from the last write before "hang" are just another "Body" frame carrying the end of the body, nothing wrong here.

The spurious mio events always comes 1 sec after we finished sending the payload (the "readable" mio events appears to take the same time when it arrives "normally"). It is indeed not because of the heartbeat timeout.

The last 8 bytes we write after the 59s "hang" are indeed the 8 bytes for an heartbeat.

Might be a coincidence, but the "truncated" write before the last one was some vectored write attempt. Running it again to see if it's just a coincidence of it it might be related.

After that I'll re run considering all events we get for the socket tells us it's readable, too see if the spurious mio event isn't in fact spurious but just doesn't have the readable flag it should.

**Comment by Keruspe** [comment]
Workaround available in bdd89d935837f48642d5c4b8765f16a8f88d7053

**Comment by jothan** [comment]
With the latest lapin, with large messages in release mode, I'm getting total stalls. No heartbeats are reported and everything grinds to a halt.

**Comment by jothan** [comment]
Strangely, as soon as the message is done submitting, the connection goes away in the RabbitMQ management interface.

**Comment by Keruspe** [comment]
With 0.39.3? (I'm currently doing lots of tests with the master branch, will do some more 0.39 tests after those)

**Comment by jothan** [comment]
No, this is with the latest master.

**Comment by Keruspe** [comment]
Master is known to have issues right now, which I'm currently debugging

**Comment by jothan** [comment]
0.39 is indeed fine. I was eagerly using the tokio executor feature.

**Comment by Keruspe** [comment]
@jothan I'm in the process of testing whether we now can revert the workaround as the underlying issues should be fixed. Just released 1.0.0-alpha3

**Comment by Keruspe** [comment]
So, latest alpha is still not enough to be able to revert the workaround. Didn't check for trace logs to see if they look the same yet though, but it's definitely better anyways as over a batch of 1000 messsages, only 4 were affected.

Will leave the workaround in place until I can send 10k without any hit

**Comment by Keruspe** [comment]
Thanks to the new Reactor, I was able to reproduce way more easily the problem (half the messages hit it when reverting the mio workaround.

After adding some debug logs to rustls when it reads or writes using the underlying socket, I noticed that all the data wasn't properly flushed, thus causing the delay. Indeed, when flushing the stream, we hit a WOULDBLOCK, but wouldn't try to write anything afterward as we didn't have any new data, all of it being in rustls internals. We now flush everytime the socket is writable.

I have one last 10k messages test running, once 100% validated I'll finally close this.
